### ISSUE TITLE
CNVKit crashes in segmentation

### ISSUE THREAD

**Issue by PeterUlz** [chapmanb/bcbio-nextgen#1232]
Hi,

I've got an issue using CNVkit as an SVcaller in tumor-only analysis.
I've created a sample yaml file including:
-) alignment with BWA
-) mark duplicates
-) realigning
-) recalibration
-) Mutect SNP caller
-) CNVkit and lumpy as SVCallers
-) a bed-file containing variant regions
-) reference hg19

While alignment runs smoothly, the program crashes in the segmentation step of CNVkit of the first sample.
The error message is:

```
$ Rscript /tmp/tmpkbRKLo

PSCBS v0.60.0 (2015-11-17) successfully loaded. See ?PSCBS for help.

Attache Paket: ‘PSCBS’

The following objects are masked from ‘package:base’:

    append, load

Loading probe coverages into a data frame
Pre-processing the probe data for segmentation
Segmenting the probe data
Error in segment(list(chrom = c(1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L, 1L,  : 
  all weights should be positive
Calls: segmentByCBS ... as.list -> as.list.listenv -> mget -> value -> value.Future
Execution halted

' returned non-zero exit status 1
Traceback (most recent call last):
```

I've tried running it in GRCh37 and by removing the "chr" from the BED-file.
I'm probably doing something wrong, but I can't figure it out.
Have you got any idea?
Best,
Peter

**Comment by chapmanb** [comment]
Peter;
Sorry about the issue. I'm not sure what is going wrong and will have to page @etal as he has great ideas about debugging CNVkit problems. Eric, have you seen this before? Do you have any clues about the best way to proceed? Thanks much.

**Comment by etal** [comment]
I haven't seen this happen before, but CNVkit assigns weights in the range [0, 1] to bins before segmentation, and it looks like here CNVkit is unexpectedly assigning weight below 0 to at least one bin. That could be triggered by something unusual in the input: huge, tiny, or negative target sizes, huge/tiny/negative read length, maybe very noisy input samples.

I can fix CNVkit to prevent negative weights, and will do so asap. To work around the issue without updating CNVkit, you could:
- Check your parameters/configuration vs. the potential triggers I listed above
- Edit the (tab-separated) .cnr files generated so far to remove the "weight" column or delete any rows where the "weight" value is zero
- Skip CNVkit in your pipeline, for now

**Comment by etal** [comment]
It looks like CNVkit already clips the "weight" array to ensure values are all above 0. Something else is happening.

Could you check the generated .cnr file to see which / how many rows have a negative or zero "weight" value, if any? If you could post or send me the `*.targetcoverage.cnn`, `*.antitargetcoverage.cnn` and `cnv_reference.cnn` files that were used to generate the problematic ".cnr" file (for one sample), I could try to replicate and diagnose the issue.

 @chapmanb -is the .cnr file passing through bedtools or any function that might alter the weight column or truncate it to an integer value?

**Comment by chapmanb** [comment]
Eric;
Thanks so much for the help with debugging. From bcbio's side, I don't think we are doing any extra processing of the cnrs. They run through `cnvkit fix` and then get passed to `cnvkit segment` directly

https://github.com/chapmanb/bcbio-nextgen/blob/489c3753f239c58f43ddf5ceecd91168cb7b5642/bcbio/structural/cnvkit.py#L157

If anything looks off in the bcbio integration I'd be happy to adjust. Thanks again.

**Comment by PeterUlz** [comment]
Hi Guys,
Thanks a lot for the help.
Indeed it seems that more than half of all the weights in the .cnr file are zero. Upon close inspection it seem that everything having "Background" in the gene field got weights of zero.

Should I upload those files (*.cnn) here? 

The commands executed from BCBio were these ( I edited file paths for clarity):
```
[2016-02-18T12:59Z] ./bcbio/anaconda/bin/bedtools merge  -i ./Test/CTC_panel/work/bedprep/Graz_NRCC_Nano_Adjacent_v2_Regions_Export.bed > ./Test/CTC_panel/work/structural/P106_1/cnvkit/tx/tmpKPu10E/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged.bed
[2016-02-18T12:59Z] cat ./Test/CTC_panel/work/structural/P106_1/cnvkit/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged.bed  | ./bcbio/anaconda/bin/pbgzip -n 10  -c > ./Test/CTC_panel/work/structural/P106_1/cnvkit/tx/tmpGlRXy2/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged.bed.gz
[2016-02-18T12:59Z] ./bcbio/anaconda/bin/tabix -f -p bed ./Test/CTC_panel/work/structural/P106_1/cnvkit/tx/tmp5Ps1cF/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged.bed.gz
[2016-02-18T12:59Z] ./bcbio/anaconda/bin/cnvkit.py access ./bcbio/genomes/Hsapiens/hg19/seq/hg19.fa -s 10000 -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/tx/tmpp9TmiQ/hg19-access.bed
[2016-02-18T13:00Z] sort -V -k1,1 -k2,2n ./Test/CTC_panel/work/structural/P106_1/cnvkit/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged.bed | bedtools closest -g <(cut -f1,2 ./bcbio/genomes/Hsapiens/hg19/seq/hg19.fa.fai | sort -V -k1,1 -k2,2n) -d -t all -a - -b <(sort -V -k1,1 -k2,2n ./bcbio/genomes/Hsapiens/hg19/rnaseq/ref-transcripts.bed) | awk -F$'\t' -v OFS='\t' '{if ($NF > 10000) $7 = "."} {print}' | cut -f 1-8 | bedtools merge -i - -c 7 -o distinct -delim ',' > ./Test/CTC_panel/work/structural/P106_1/cnvkit/tx/tmptnEW85/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated.bed
[2016-02-18T13:00Z] ./bin/cnvkit.py target ./Test/CTC_panel/work/structural/P106_1/cnvkit/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated.bed --split -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/tx/tmpSNLkml/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated.target.bed
[2016-02-18T13:00Z] ./bin/cnvkit.py antitarget -g ./Test/CTC_panel/work/structural/P106_1/cnvkit/hg19-access.bed ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated.target.bed -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/tx/tmp8RRK7o/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated.antitarget.bed
[2016-02-18T13:00Z] ./bcbio/anaconda/bin/cnvkit.py coverage ./Test/CTC_panel/work/align/P106_1/3_2016-02-15_CTC_Panel-sort.bam ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated-0.target.bed -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/tx/tmp5MVmDM/3_2016-02-15_CTC_Panel-sort-0.targetcoverage.cnn
[2016-02-18T13:04Z] ./bcbio/anaconda/bin/cnvkit.py coverage ./Test/CTC_panel/work/align/P106_1/3_2016-02-15_CTC_Panel-sort.bam ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated-0.antitarget.bed -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/tx/tmpIU5fbT/3_2016-02-15_CTC_Panel-sort-0.antitargetcoverage.cnn
[2016-02-18T13:04Z] ./bcbio/anaconda/bin/cnvkit.py reference -f ./bcbio/genomes/Hsapiens/hg19/seq/hg19.fa -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/tx/tmpTHtS6B/flat_background.cnn -t ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated.target.bed -a ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/Graz_NRCC_Nano_Adjacent_v2_Regions_Export-merged-annotated.antitarget.bed
[2016-02-18T13:05Z] ./bcbio/anaconda/bin/cnvkit.py fix -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/tx/tmpXIC50w/3_2016-02-15_CTC_Panel-sort.cnr ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/3_2016-02-15_CTC_Panel-sort.targetcoverage.cnn ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/3_2016-02-15_CTC_Panel-sort.antitargetcoverage.cnn ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/flat_background.cnn
[2016-02-18T13:05Z] unset R_HOME && export PATH=./bcbio/anaconda/bin:$PATH && ./bcbio/anaconda/bin/cnvkit.py segment -o ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/tx/tmpX7FfRG/3_2016-02-15_CTC_Panel-sort.cns --rlibpath ./tools/lib/R/site-library ./Test/CTC_panel/work/structural/P106_1/cnvkit/raw/3_2016-02-15_CTC_Panel-sort.cnr
```
Do you want me to upload the CNN files here?
Best
Peter

**Comment by chapmanb** [comment]
Peter;
Thanks much for confirming the issue. If you could post the .cnn files as separate gists (https://gist.github.com/) and link to them here that would help a lot. Thanks again for helping to debug this.

**Comment by PeterUlz** [comment]
CNR file: https://gist.github.com/PeterUlz/748a592a75d77b4eccbd
Antitarget CNN file: https://gist.github.com/PeterUlz/0741d1a9dd4d81e60eb2
Target CNN file: https://gist.github.com/PeterUlz/1d90d86de7f5099cabe9
Flat Background CNN: https://gist.github.com/PeterUlz/21c3cd715858bb9016fd

BCBio config file: https://gist.github.com/PeterUlz/8f6761729ae4336c718e

If you need anything more for debugging just tell me.

**Comment by etal** [comment]
In the target .cnn file, the majority of bins had no coverage at all in the BAM file (according to CNVkit) and received the default minimum log2 value of -20. Is it possible that this sample is bad, the target panel/BED is a mismatch, or something else went wrong in your pipeline?

CNVkit compares the variability (interquartile range of log2 values) of the target and antitarget bins, and calculates a ratio and downweights the more variable set of bins in proportion. Since the top and bottom quartiles of the target bins are both -20, the IQR is zero, and the ratio of antitarget-to-target IQRs is infinity. The antitarget weights are then divided by infinity, leaving them all zero. (I did not anticipate any of that happening.)

I've fixed CNVkit to prevent the crash, and successfully segmented your .cnr file, but the resulting copy number profile does not look very useful because of the underlying data issue. If you figure out what happened, and if this is something you expect to see in practice occasionally, I'd appreciate hearing about the root cause so I can try to make CNVkit handle it better.
